phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java

```diff
@@ -0,0 +1,280 @@
+            for (Annotation annotation : constructor.getAnnotations()) {
+                if (annotation.annotationType() == JsonCreator.class) {
+                    return Optional.of(constructor);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<Constructor<?>> getConstructorHavingJsonBuilder(Class<?> clazz) {
+        for (Constructor<?> constructor : clazz.getDeclaredConstructors()) {
+            final List<Class<?>> parameterTypes = Arrays.stream(constructor.getParameterTypes())
+                    .filter(s -> s.getName().contains("Builder")).collect(Collectors.toList());
+
+            if (parameterTypes.size() > 0) {
+                final Annotation[] annotations = parameterTypes.get(0).getAnnotations();
+                if (annotations.length > 0 && JsonPOJOBuilder.class.equals(annotations[0].annotationType())) {
+                    return Optional.of(constructor);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static List<DTOContainer<?>> findAllDTOs() throws ClassNotFoundException {
+        final List<DTOContainer<?>> fondJsonCreatorClasses = new ArrayList<>();
+
+        final ClassPathScanningCandidateComponentProvider scanner = new ClassPathScanningCandidateComponentProvider(
+                true);
+        scanner.addIncludeFilter((metadataReader, metadataReaderFactory) -> true);
+
+        final Stream<String> classNamesStream = scanner.findCandidateComponents("org.phoenicis").stream()
+                .map(BeanDefinition::getBeanClassName);
+        final Iterable<String> classNames = classNamesStream::iterator;
+
+        for (String className : classNames) {
+            final Class<?> clazz = Class.forName(className);
+            if (isJsonClass(clazz)) {
+                fondJsonCreatorClasses.add(new DTOContainer<>(clazz));
+            }
+        }
+
+        return fondJsonCreatorClasses;
+    }
+
+    private static boolean isJsonClass(Class<?> clazz) {
+        return getConstructorAnnotatedWithJsonCreator(clazz).isPresent()
+                || getConstructorHavingJsonBuilder(clazz).isPresent();
+    }
+
+    @Test
+    public void testConstructor() throws ReflectiveOperationException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        assertNotNull(dto);
+    }
+
+    @Test
+    public void testHashCodes() throws ReflectiveOperationException, URISyntaxException {
+        final Object dto1 = newDTOInstance(DTOContainer.getClazz());
+        final Object dto2 = newDTOInstance(DTOContainer.getClazz());
+        assertEquals(dto1, dto2);
+        assertEquals(dto1.hashCode(), dto2.hashCode());
+    }
+
+    @Test
+    public void testEqualsTrivialCases() throws ReflectiveOperationException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        assertTrue(dto.equals(dto));
```

Comment:
No fondamental differences but in this particular case, we are testing the equals method. So we really want to call it explicitely